File: 1999/2/6-2.py
#!/usr/bin/env python

import sys

for line in sys.stdin:
    numcoins = int(line)
    p1 = []
    p2 = []
    worst = numcoins * (numcoins + 1) / 4
    best = numcoins * (numcoins + 1) / 2 - worst
    sum1 = 0
    sum2 = 0
    for i in xrange(numcoins,0,-1):
        if sum1 + i <= best:
            sum1 += i
            p1.append(i)
        else:
            sum2 += i
            p2.append(i)
    p1.reverse()
    p2.reverse()
    print
    print "Number of coins:", numcoins
    print "Pile 1 contains",
    print ", ".join(map(str, p1)),
    print "with value %s." % sum1
    print "Pile 2 contains",
    print ", ".join(map(str, p2)),
    print "with value %s." % sum2

File: 1999/2/6-2.cpp
#include <iostream>
#include <fstream>
#include <deque>
using namespace std;

void main()
{
	ifstream in("prob2.in");
	int numcoins;
	while (in >> numcoins)
	{
		deque<int> p1, p2;
		int worst = (numcoins * (numcoins + 1) / 4),
			best = (numcoins * (numcoins + 1) / 2) - worst,
			sum1 = 0, sum2 = 0;
		for (int i = numcoins; i > 0; i--)
		{
			if ((sum1 + i) <= best)
			{
				sum1 += i;
				p1.push_front(i);
			}
			else
			{
				sum2 += i;
				p2.push_front(i);
			}
		}
		cout << "Number of coins: " << numcoins << endl << "Pile 1 contains ";
		for (i = 0; i < p1.size(); i++)
			cout << p1[i] << ((i == p1.size() - 1) ? " ":", ");
		cout << "with value " << sum1 << "." << endl << "Pile 2 contiains ";
		for (i = 0; i < p2.size(); i++)
			cout << p2[i] << ((i == p2.size() - 1) ? " ":", ");
		cout << "with value " << sum2 << "." << endl;
	}
}  
